examples/server/server.cpp

```diff
@@ -49,347 +119,309 @@ struct llama_server_context
+    const float mirostat_tau = params.mirostat_tau;
+    const float mirostat_eta = params.mirostat_eta;
+    const bool penalize_nl = params.penalize_nl;
+    llama_token id = 0;
+    {
+      auto *logits = llama_get_logits(ctx);
+      auto n_vocab = llama_n_vocab(ctx);
 
-        llama_token_data_array candidates_p = {candidates.data(), candidates.size(), false};
-
-        // Apply penalties
-        float nl_logit = logits[llama_token_nl()];
-        auto last_n_repeat = std::min(std::min((int)last_n_tokens.size(), repeat_last_n), params.n_ctx);
-        llama_sample_repetition_penalty(ctx, &candidates_p,
-                                        last_n_tokens.data() + last_n_tokens.size() - last_n_repeat,
-                                        last_n_repeat, repeat_penalty);
-        llama_sample_frequency_and_presence_penalties(ctx, &candidates_p,
-                                                      last_n_tokens.data() + last_n_tokens.size() - last_n_repeat,
-                                                      last_n_repeat, alpha_frequency, alpha_presence);
-        if (!penalize_nl)
-        {
-          logits[llama_token_nl()] = nl_logit;
-        }
+      // Apply params.logit_bias map
+      for (const auto &it : params.logit_bias) {
+        logits[it.first] += it.second;
+      }
 
-        if (temp <= 0)
-        {
-          // Greedy sampling
-          id = llama_sample_token_greedy(ctx, &candidates_p);
-        }
-        else
-        {
-          if (mirostat == 1)
-          {
-            static float mirostat_mu = 2.0f * mirostat_tau;
-            const int mirostat_m = 100;
-            llama_sample_temperature(ctx, &candidates_p, temp);
-            id = llama_sample_token_mirostat(ctx, &candidates_p, mirostat_tau, mirostat_eta, mirostat_m, &mirostat_mu);
-          }
-          else if (mirostat == 2)
-          {
-            static float mirostat_mu = 2.0f * mirostat_tau;
-            llama_sample_temperature(ctx, &candidates_p, temp);
-            id = llama_sample_token_mirostat_v2(ctx, &candidates_p, mirostat_tau, mirostat_eta, &mirostat_mu);
-          }
-          else
-          {
-            // Temperature sampling
-            llama_sample_tail_free(ctx, &candidates_p, tfs_z, 1);
-            llama_sample_typical(ctx, &candidates_p, typical_p, 1);
-            llama_sample_top_p(ctx, &candidates_p, top_p, 1);
-            llama_sample_temperature(ctx, &candidates_p, temp);
-            id = llama_sample_token(ctx, &candidates_p);
-          }
-        }
-        last_n_tokens.erase(last_n_tokens.begin());
-        last_n_tokens.push_back(id);
-        processed_tokens.push_back(id);
-        num_tokens_predicted++;
+      std::vector<llama_token_data> candidates;
+      candidates.reserve(n_vocab);
+      for (llama_token token_id = 0; token_id < n_vocab; token_id++)
+      {
+        candidates.emplace_back(llama_token_data{token_id, logits[token_id], 0.0f});
```

Comment:
`emplace_back()` does forward the parameters.
if it for some reason does not work, use normal `push_back()`, it is equivalent when construct the object before hand.

Reply:
Weirdly, I get a build error when I strip the constructor out which is why I asked without looking into it more. I've looked into the build error, and I'm still confused.

`'llama_token_data::llama_token_data': no overloaded function takes 3 arguments`

That's with VS2022. Not sure what the issue is since the struct definitely takes those 3 arguments.

Reply:
oh yea, we had that issue before. Just change it to `push_back()` so we don't pretend to be better than we are. :)

and wait for a better world...

Reply:
push_back() gives the same error. The world may be beyond hope. For VS2022 anyway.

Reply:
yes, but we don't make it look like it is more performant.

Reply:
Do we need to push the objects? `n_vocab` is fixed, we could just use array indexing?

Reply:
Ah, I misunderstood. So throw the constructor back in, but we use push_back() to say we're sad and hope for brighter days ahead.

@SlyEcho Do we have a constructed set of llama_token_data to index from as is? It's all running up after the logit_bias changes so I assume that's why it's rebuilding the array. Or are you saying just build them one to one, like 

`candidates[token_id] = llama_token_data{token_id, logits[token_id], 0.0f}`?

Reply:
I vote if it's good enough for main.cpp and isn't adding significant runtime to a given token response overall, it's probably fine. And I agree. Probably better to make it recognizable for anyone else who comes wandering by if new sampling methods get added or (ideally) the sampling code gets rolled out somewhere else.